### PR TITLE
Trac3528

### PR DESCRIPTION
**Pull request by bennyfio-ebi**
Please also verify that acceptance tests (/test/acceptance/rest) are successfull because they haven't been yet automated in the bamboo CI process.
The tests require Atlas running on localhost.
They can be executed from the IDE or at command line specifying the acceptance profile:
mvn test -P acceptance

### CODE REVIEW

File: atlas-dao/src/main/java/uk/ac/ebi/gxa/dao/ExperimentDAO.java [hunk 1]
```diff
@@ -1,14 +1,19 @@
 package uk.ac.ebi.gxa.dao;
 
+import com.google.common.base.Function;
```
Comment (nsklyar) [inline atlas-dao/src/main/java/uk/ac/ebi/gxa/dao/ExperimentDAO.java:3]: unused imports.
use [cmd]-[alt]-[o] to optimize imports.

File: atlas-dao/src/main/java/uk/ac/ebi/gxa/dao/ExperimentDAO.java [hunk 2]
```diff
@@ -47,6 +52,29 @@ public ExperimentDAO(SessionFactory sessionFactory) {
         return template.find("select e from Experiment e left join e.assays a left join a.properties p where p.propertyValue.property.name = ?  and p.propertyValue.value = ?", propertyName, propertyValue);
     }
 
+    @SuppressWarnings("unchecked")
+    public List<Experiment> getExperimentsByProperties(ApiProperty[] properties) {
```
Comment (nsklyar) [inline atlas-dao/src/main/java/uk/ac/ebi/gxa/dao/ExperimentDAO.java:56]: here we are introducing dependencies to classes from model/api in dao class, though I'm not sure I see now better way.

File: atlas-model/src/main/java/uk/ac/ebi/microarray/atlas/model/OntologyTerm.java [hunk 3]
```diff
@@ -70,6 +72,19 @@ public void setTerm(String term) {
...
+    }
+
+    @Override
+    public boolean equals(Object other){
+        if(accession != null && other !=null && other instanceof OntologyTerm){
+            return Objects.equal(accession, ((OntologyTerm) other).getAccession());
+        }
+        return false;
+    }
+
```
Comment (nsklyar) [inline atlas-model/src/main/java/uk/ac/ebi/microarray/atlas/model/OntologyTerm.java:87]: why not to have hashCode() and equals() like they are generated by IntelliJ?

File: atlas-model/src/test/java/uk/ac/ebi/microarray/atlas/api/ApiPropertyTest.java [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+
+    @Test
+    public void constructorThatAcceptsPropertyValueAndTermsShouldSimplySetThem(){
+
+        //when
+        ApiProperty apiProperty = new ApiProperty(apiPropertyValueMock, ontologyTermsMock);
+
+        //then
+        assertThat(apiProperty.getPropertyValue(), equalTo(apiPropertyValueMock));
+        assertThat(apiProperty.getTerms(), equalTo(ontologyTermsMock));
```
Comment (nsklyar) [inline atlas-model/src/test/java/uk/ac/ebi/microarray/atlas/api/ApiPropertyTest.java:108]: ontologyTermsMock is empty, probably we need to test with non empty set es well.

File: atlas-model/src/test/java/uk/ac/ebi/microarray/atlas/api/ApiPropertyTest.java [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+
+        //given
+        given(assayPropertyMock.getPropertyValue()).willReturn(propertyValueMock);
+
+        //when
+        ApiProperty apiProperty = new ApiProperty(assayPropertyMock);
+
+        //then
+        assertThat(apiProperty.getPropertyValue().getValue(), equalTo(assayPropertyMock.getPropertyValue().getValue()));
+        assertThat(apiProperty.getPropertyValue().getProperty().getName(), equalTo(assayPropertyMock.getPropertyValue().getDefinition().getName())); //B: 4 level of delegation to get to a property name
```
Comment (nsklyar) [inline atlas-model/src/test/java/uk/ac/ebi/microarray/atlas/api/ApiPropertyTest.java:127]: it seems that here we are testing ApiPropertyValue constructor, rather than ApiProperty

File: atlas-model/src/main/java/uk/ac/ebi/microarray/atlas/api/ApiProperty.java [hunk 6]
```diff
@@ -43,6 +43,14 @@ public ApiProperty(final SampleProperty assayProperty) {
                         instanceTransformer(OntologyTerm.class, ApiOntologyTerm.class)));
     }
 
+    public String getName(){
+        return propertyValue.getProperty().getName();
+    }
+
+    public String getValue(){
+        return propertyValue.getValue();
```
Comment (nsklyar) [inline atlas-model/src/main/java/uk/ac/ebi/microarray/atlas/api/ApiProperty.java:51]: here and in the method above NPE might be thrown if propertyValue is not initialized. And this situation is possible since we have empty public constructor ( I'm not sure why we need )

File: atlas-model/src/test/java/uk/ac/ebi/microarray/atlas/api/ApiPropertyTest.java [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+    }
+
+    @Test
+    public void setPropertyValueShouldDoWhatItSays() throws Exception {
+
+        //when
+        subject.setPropertyValue(apiPropertyValueMock);
+
+        //then
+        assertThat(subject.getPropertyValue(), equalTo(apiPropertyValueMock));
```
Comment (nsklyar) [inline atlas-model/src/test/java/uk/ac/ebi/microarray/atlas/api/ApiPropertyTest.java:169]: do we want to test getters/setters?

File: atlas-model/src/test/java/uk/ac/ebi/microarray/atlas/model/OntologyTermEqualityAndHashCodeTest.java [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+    public void initializeOntologyTerm(){
+
+        subject = new OntologyTerm(null, ontologyMock, "TERM_VALUE", "ACCESSION_VALUE", "DESCRIPTION_VALUE");
+
+    }
+
+    @Test
+    public void hashCodeShouldBeTheSameAsAccessionHashCode() throws Exception {
+
+        assertThat(subject.hashCode(), equalTo(Objects.hashCode("ACCESSION_VALUE")));
```
Comment (nsklyar) [inline atlas-model/src/test/java/uk/ac/ebi/microarray/atlas/model/OntologyTermEqualityAndHashCodeTest.java:61]: we are testing implementation here!

File: atlas-web/pom.xml [hunk 9]
```diff
@@ -395,6 +395,44 @@
...
+            <exclusions>
+                <exclusion>
+                    <artifactId>hamcrest-library</artifactId>
+                    <groupId>org.hamcrest</groupId>
+                </exclusion>
+                <exclusion>
+                    <artifactId>hamcrest-core</artifactId>
+                    <groupId>org.hamcrest</groupId>
+                </exclusion>
+            </exclusions>
```
Comment (nsklyar) [inline atlas-web/pom.xml:416]: can we have scope "test"?

File: atlas-web/src/main/java/uk/ac/ebi/gxa/service/CurationService.java [hunk 10]
```diff
@@ -538,11 +538,88 @@ public void putAssayProperties(final String experimentAccession,
...
 
     /**
+     * Adds or updates mapping to efo terms for matching assayProperties to all assays in experiment: experimentAccession
+     *
+     * @param experimentAccession String
+     * @param properties ApiProperty[]
+     * @throws ResourceNotFoundException if experiment: experimentAccession in that experiment are not found
+     */
+    @Transactional
+    public void remapTermsOnMatchingAssayProperties(String experimentAccession, ApiProperty[] properties) throws ResourceNotFoundException {
```
Comment (nsklyar) [inline atlas-web/src/main/java/uk/ac/ebi/gxa/service/CurationService.java:548]: this and following new methods are not tested in CurationServiceTest

File: atlas-web/src/test/java/acceptance/rest/GetExperimentTestExternal.java [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+
+        expect().body("samples", not(empty()))
+            .and()
+            .body("assays", not(empty()))
+            .when().get(EXPERIMENT_URI);
+
+    }
+
+    @Test
+    public void samplesShouldContainSomeProperties() throws Exception {
```
Comment (nsklyar) [inline atlas-web/src/test/java/acceptance/rest/GetExperimentTestExternal.java:87]: this test fails

File: atlas-web/src/main/java/uk/ac/ebi/gxa/service/CurationService.java [hunk 12]
```diff
@@ -561,8 +638,86 @@ public void deleteAssayProperties(final String experimentAccession,
...
     /**
-     * @param experimentAccession
-     * @param sampleAccession
+     * Removes properties from all assays and all samples in all experiments
+     *
+     * @param properties ApiProperty[]
+     * @throws ResourceNotFoundException if experiment: experimentAccession is not found
+     */
+    @Transactional
+    public void deleteTermsFromMatchingPropertiesForAllAllExperiments(ApiProperty[] properties) throws ResourceNotFoundException {
```
Comment (nsklyar) [inline atlas-web/src/main/java/uk/ac/ebi/gxa/service/CurationService.java:647]: I guess you meant deleteTermsFromMatchingPropertiesForAllExperiments :)

File: atlas-model/src/test/java/uk/ac/ebi/microarray/atlas/api/ApiPropertyTest.java [hunk 13]
```diff
@@ -0,0 +1,223 @@
...
+    }
+
+    @Test
+    public void getPropertyValueShouldReturnNullWhenThePropertyIsNotInitialized() throws Exception {
+
+        //given
+        subject = new ApiProperty();
+
+        //when
+        ApiPropertyValue propertyValue = subject.getPropertyValue();
```
Comment (nsklyar) [inline atlas-model/src/test/java/uk/ac/ebi/microarray/atlas/api/ApiPropertyTest.java:155]: this is test

File: atlas-web/src/main/java/uk/ac/ebi/gxa/service/CurationService.java [hunk 14]
```diff
@@ -561,8 +638,86 @@ public void deleteAssayProperties(final String experimentAccession,
...
+
+            for (Experiment experiment: experimentDAO.getExperimentsByProperties(properties)) {
+
+                deleteTermsFromMatchingPropertiesForAllAssays(experiment, properties);
+                deleteTermsFromMatchingPropertiesForAllSamples(experiment, properties);
+
+            }
```
Comment (nsklyar) [inline atlas-web/src/main/java/uk/ac/ebi/gxa/service/CurationService.java:654]: do we actually need to fetch all experiments and than iterate over them? probalby we can just delete ontology terms directly from Assay/Sample properies
